### DISCUSSION TITLE
Practice mode (and maybe save points for it)

### DISCUSSION THREAD

**Discussion by BlockHead007** [ppy/osu#27867] [Ideas] [answered]
There is #20323 already, but here is a little bit different thing. It works same as in adofai (or maybe gd, dunno, it works a little bit different), supporting all rulesets, including customs.

During gameplay player can pause game process and enter Practice mode (score cannot be submitted since then, all mods are remembered). Game remembers when beatmap was paused and sets save point here. From this moment player beats the rest section until he decides to try again: pressing Quick retry brings player back to save point. Also, he can make new save points at different spots. Reaching next save point means that player can't go back to previous and he needs to seek for it manually from menu, from now on every retry will bring player to the new point. At the exit of this mode beatmap will be fully restarted, can be played to the end and score will be submitted.

Save points may be unnecessary because adofai is about memorizing charts and here in osu you mostly should just learn reading beatmaps, but that practice mode thing would be a very nice addition.

**Comment by Walavouchey** [comment, marked as answer]
to group things together it'd be better if you posted in the existing thread instead. it's better than 10 threads with 10 variations of one feature

**Reply by Joehuu** [reply]
There's also https://github.com/ppy/osu/discussions/13013 with more upvotes / comments.
